File: 24-Template-Method/src/go/beverage.go
package main

import "fmt"

// Beverage is the interface of the VARYING steps. Go has no inheritance, so
// the template method is a free function (PrepareRecipe) that takes this
// interface and calls the steps in the fixed order.
type Beverage interface {
	Brew()
	AddCondiments()
	WantsCondiments() bool
}

// Invariant steps — shared by all beverages.
func boilWater() { fmt.Println("  Boiling water") }
func pourInCup() { fmt.Println("  Pouring into cup") }

// PrepareRecipe is the template method: the fixed skeleton and ordering.
func PrepareRecipe(b Beverage) {
	boilWater()
	b.Brew()
	pourInCup()
	if b.WantsCondiments() {
		b.AddCondiments()
	}
}

File: 24-Template-Method/src/rust/beverage.rs
/// A trait with a default method IS the Template Method in Rust: the default
/// `prepare_recipe` fixes the skeleton and calls the required steps and the
/// `wants_condiments` hook. Implementors provide only the varying steps.
pub trait CaffeineBeverage {
    // Varying steps — implementors must supply these.
    fn brew(&self);
    fn add_condiments(&self);

    // Hook — default a type may override.
    fn wants_condiments(&self) -> bool {
        true
    }

    // Template method — the fixed skeleton and ordering.
    fn prepare_recipe(&self) {
        boil_water();
        self.brew();
        pour_in_cup();
        if self.wants_condiments() {
            self.add_condiments();
        }
    }
}

// Invariant steps — shared by all beverages.
fn boil_water() {
    println!("  Boiling water");
}

fn pour_in_cup() {
    println!("  Pouring into cup");
}
